plutus-core/plutus-core/src/PlutusCore/ParserCommon.hs

```diff
@@ -0,0 +1,202 @@
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c = isAlphaNum c || c == '_' || c == '\''
+
+-- | Create a parser that matches the input word and returns its source position.
+-- This is for attaching source positions to parsed terms/programs.
+-- getSourcePos is not cheap, don't call it on matching of every token.
+reservedWord ::
+    -- | The word to match
+    T.Text -> Parser SourcePos
+reservedWord w = lexeme $ try $ do
+    p <- getSourcePos
+    void $ string w
+    notFollowedBy (satisfy isIdentifierChar)
+    return p
+
+builtinFunction :: (Bounded fun, Enum fun, Pretty fun) => Parser fun
+builtinFunction = lexeme $ choice $ map parseBuiltin [minBound .. maxBound]
+    where parseBuiltin builtin = try $ string (display builtin) >> pure builtin
+
+version :: Parser (PLC.Version SourcePos)
+version = lexeme $ do
+    p <- getSourcePos
+    x <- Lex.decimal
+    void $ char '.'
+    y <- Lex.decimal
+    void $ char '.'
+    PLC.Version p x y <$> Lex.decimal
+
+name :: Parser PLC.Name
+name = lexeme $ try $ do
+    void $ lookAhead letterChar
+    str <- takeWhileP (Just "identifier") isIdentifierChar
+    PLC.Name str <$> intern str
+
+tyName :: Parser PLC.TyName
+tyName = PLC.TyName <$> name
+
+-- | Turn a parser that can succeed without consuming any input into one that fails in this case.
+enforce :: Parser a -> Parser a
+enforce p = do
+    (input, x) <- match p
+    guard . not $ T.null input
+    pure x
+
+-- | Consume a chunk of text and either stop on whitespace (and consume it) or on a \')\' (without
+-- consuming it). We need this for collecting a @Text@ to pass it to 'PLC.parse' in order to
+-- parse a built-in type or a constant. This is neither efficient (because of @manyTill anySingle@),
+-- nor future-proof (what if some future built-in type has parens in its syntax?). A good way of
+-- resolving this would be to turn 'PLC.parse' into a proper parser rather than just a function
+-- from @Text@, but PLC's parser also uses 'PLC.parse' and it's currently not @megaparsec@-based,
+-- and so mixing two distinct parsing machines doesn't sound too exciting.
+--
+-- Note that this also fails on @(con string \"yes (no)\")@ as well as @con unit ()@, so it really
+-- should be fixed somehow.
+closedChunk :: Parser T.Text
+closedChunk = T.pack <$> manyTill anySingle end where
+    end = enforce whitespace <|> void (lookAhead $ char ')')
+
+-- | Parse a type tag by feeding the output of 'closedChunk' to 'PLC.parse'.
+builtinTypeTag
+    :: PLC.Parsable (PLC.SomeTypeIn (PLC.Kinded uni))
+    => Parser (PLC.SomeTypeIn (PLC.Kinded uni))
+builtinTypeTag = do
+    uniText <- closedChunk
+    case PLC.parse uniText of
+        Nothing  -> customFailure $ UnknownBuiltinType uniText
+        Just uni -> pure uni
+
+-- | Parse a constant by parsing a type tag first and using the type-specific parser of constants.
```

Comment:
I think this'll have problems with some built-in constants like strings and characters.  The existing parser has special cases involving complicated regular expressions to deal with those (see `Lexer.x`), but things got more complicated recently when @effectfully added built-in lists and pairs that can have other constants nested inside them. I don't think we've quite worked out how to deal with that yet. We're probably still going to need special parsers for things like quoted strings that can contain escape sequences.

Reply:
Noted and added to comments. Hopefully I can solve some of these problems in the next PR.